### DISCUSSION TITLE
Sumatra's Bookmarks

### DISCUSSION THREAD

**Discussion by PAL-Thailand** [sumatrapdfreader/sumatrapdf#4674] [General]
I've been using the SumatraPDF Reader for quite some time and find it to be a well thought-out program. I am especially fond of the Bookmarks which show up in the left panel.  Recently, I used SumatraPDF to open "PowerShell-scripting-powershell-7.4.pdf". The file is HUGE -- like 83.4 MB huge.  SumatraPDF handles the file very well in the right side viewing  area.  But what pleased me more was the Bookmarks panel that provides two types of bookmarks for the PDF above, i.e. collapsed or expanded.  Super neat!

It is so neat that I have been trying to copy or extract the complete contents of the Bookmarks. In this endeavor, I have failed! I found no way to for SumatraPDF to allow me the copy or extract just what appears in the left panel where the Bookmarks are shown.

Is there any way to extract the complete expanded contents of the 
Bookmarks without having to rely on third party apps that will only do a pittance of the huge PDF I am reading bit by bit.  I would really to have a print-out of what the Bookmarks show which I equivocate as the PDF's Table of Contents (TOC) as a reference sheet of the PDF's contents at my workdesk.

I am supposing that the creators of SumatraPDF may have been asked the same question that I am bringing up today.  Any and all comments, suggestions, tips, work-arounds to extract whatever information about the makeup of the pdfs it can read.  I'd be most happy to hear from others who may have similar wishes.  Thanks for reading.  Peter in Thailand

**Comment by GitHubRulesOK** [comment]
There WAS an experimental set of versions some time ago that worked with actual and "virtual" bookmarks (without edit the source file) but you would need to built one of those, which were abandoned as "problematic". Basically there are many ways bookmarks can be managed but that also means a high quantity of code for their management, and SumatraPDF is not designed as an editor simply a reader.

On a personal level I have tried to produce a simple "add" new bookmark without edits (in theory they are just like any other annotation) but needs be uses 3rd party command line and thus can only be of very limited value. However the core may possibly be adapted to export only.
![image](https://github.com/user-attachments/assets/c5373016-7039-4dde-b7ed-c98de89cc26d)
all it does is call a cmd file to edit the current file with the current page number. `Local\SumatraPDF\plus\AddBookmark.cmd" "%1" %p
		Name = &Bookmark this PDF page` and as it currently stands is unlikely to work with files over 32 MB ? due to "file locking"


By far the simplest is use a dedicated extract as text, "list, edit, add and **replace the lot**" method using notepad. The best lightest & powerful for the task is possibly Coherent `cpdf` but there are others also available as one exe that can export easily.

**Comment by PAL-Thailand** [comment]
Many thanks for your input. Having read it, it dawned on me that the word "Bookmarks", which can be made visible by clicking the 3 small horizontal lines at the top left of the app, then select 'View' > 'Bookmarks', may have different meanings for different people.  The 3,401-page PDF does not have a TOC at the beginning of the PDF document I am dealing with.  Somehow, SumatraPDF has created a compilation of all the headings and sub-headings from the contents of the document AND and places it in the left panel entitled "Bookmarks". 

My desire/wish/hope is to extract the contents what SumatraPDF has created and displayed in the left side panel as can be seen in the .jpg file I hope to attach to my comment.  I do not want to extract one or many pages from the original PDF.  I do not want to create, change or edit a bookmark.  I do not have the proper nomenclature for the various sections, screens, windows that Sumatra PDF has and it may be this deficiency on my part to correctly point to various areas or parts of the SumatraPDF application.  Please advise how I can attach a .jpg file to this comment.

**Reply by GitHubRulesOK** [reply]
They are commonly referred to as "bookmarks" but are a set of Document Outline entries nested, often randomly, throughout a PDF when required thus not all in one place for extraction. the entries look something like when I add one to each page.

![image](https://github.com/user-attachments/assets/2fb5c365-868f-4e62-9577-c0d7a9171bd1)


one entry will be perhaps
```
9 0 obj
<</Type/Page/Parent 1 0 R/Resources<</Font<</F1<</Type/Font/Subtype/Type1/BaseFont/Helvetica>>>>>>/Contents 5 0 R/Annots[8 0 R]>>
endobj
8 0 obj
<</Type/Annot/Subtype/FreeText/Rect[270.62574 601.87429 471.62574 702.87429]/CL[0 792 271.12574 689.99649]/BS<</Type/Border/W 1>>/DA(/Helv 12 Tf 0 0 0 rg)/P 9 0 R/F 4/M(D:20241117145038Z)/Contents(This is a text...)/RD[0 0 0 0]/AP<</N 3 0 R>>>>
endobj
4 0 obj
<</Parent 2 0 R/Dest[9 0 R/FitH 842]/Title(Hello World \(Page 5\))>>
endobj
2 0 obj
<</First 4 0 R/Last 4 0 R>>
endobj
```

but that is based on a simple list so the export is
```
0 "Page 1" 1 "[1/FitH 842]"
0 "Page 2" 2 "[2/FitH 842]"
0 "Page Third" 3 "[3/FitH 842]"
0  "Hello World (Page 5)" 5 "[5 /FitH 842]"
```

Available as AddBookmark.cmd
As mentioned that is using cpdf and downloads the 32bit version on very first run then uses it to ADD a bookmark but if you know how to edit can be altered to just do the EXPORT.
all I did was REM out the deletion line (third line in this group)
```
:eof
REM clean-up
chcp %_codepage% >nul
REM del /q "%temp%\bookmarks-out.txt"
exit /b
```

I have uploaded the modified list only version so see if it works for you

https://github.com/GitHubRulesOK/SumatraPDF-Plus/blob/master/Plus/List-Bookmarks.cmd

To run it you need in Advanced Options an entry like
```
ExternalViewers [
	[
		CommandLine = "C:\Users\ PUT your user name here \AppData\Local\SumatraPDF\plus\List-Bookmarks.cmd" "%1" 
		Name = Export all &Bookmarks for this PDF as filename.bkm
		Filter = *.pdf
		Key = B
	]
]
```
